Sprint 15

Aleksandr0406 opened the pull request:
1. На этот экран добавлены коллекции для выбора эмоджи и цвета; в в Tracker добавлены кнопки для перехода на создание категории и расписания
2. Реализован слой работы с базой данных с использованием абстракций
3. Дизайн проекта соответствует дизайну макета в Figma, включая все элементы, шрифты, цвета и отступы

Aleksandr0406 committed:
refactor: удалить неиспользуемые принты

Aleksandr0406 committed:
feat: добавить поддержку core data в проект

Aleksandr0406 committed:
refactor: общий рефакт

Aleksandr0406 committed:
refactor: перенести файлы по папкам

Aleksandr0406 committed:
refactor: общий рефакт

Aleksandr0406 committed:
feat: добавить возможность сохранения выбранного эмоджи и цвета трекера

Aleksandr0406 committed:
refactor: общий рефакт

Aleksandr0406 committed:
feat: добавить возможность ячейки трекера показывать всю выбранную информацию

Aleksandr0406 committed:
feat: добавить сохранение трекеров в базу данных через CoreData

Andreynnt commented on Tracker/CoreData/StoreClasses/TrackerCategoryStore.swift:
```diff
@@ -0,0 +1,179 @@
+    
+    convenience override init() {
+        let context = (UIApplication.shared.delegate as! AppDelegate).persistentContainer.viewContext
+        try! self.init(context: context)
```
Всегда стоит избегать force unwrap операций, это может крашнуть приложение. Давай поправим тут и в остальных местах

Aleksandr0406 replied:
поправил везде кроме мест, где встречается  try! self.init(context: context). Чем бы не распаковывал приводит к ошибке- 'self.init' isn't called on all paths before returning from initializer. И исправить ее пока не получилось. Насколько критично оставить здесь форс анрэп?

Andreynnt commented on Tracker/DataProvider/DataProvider.swift:
```diff
@@ -0,0 +1,58 @@
+    private let trackerRecordStore: TrackerRecordStore = TrackerRecordStore()
+    
+    private var context: NSManagedObjectContext? {
+        return (UIApplication.shared.delegate as! AppDelegate).persistentContainer.viewContext
```
В одностросных функциях return можно не писать. От force cast-a стоит отказаться

Andreynnt commented on Tracker/Helpers/TrackerScheduleValueTransformer.swift:
```diff
@@ -0,0 +1,31 @@
+    
+    override func reverseTransformedValue(_ value: Any?) -> Any? {
+        guard let data = value as? NSData else { return nil }
+        return try? JSONDecoder().decode([Int].self, from: data as Data)
```
Давай декодера положим в свойства класса, чтобы каждый раз его не создавать

Andreynnt commented on Tracker/NewHabbitViewController/NewHabitViewController.swift:
```diff
@@ -390,8 +501,8 @@ extension NewHabitViewController: UITableViewDelegate {
             viewcontroller.onDoneButtonTapped = { days in
```
Тут потерян weak, нужно добавить

Andreynnt commented on Tracker/NewHabbitViewController/NewHabitViewController.swift:
```diff
@@ -374,12 +462,35 @@ extension NewHabitViewController: UICollectionViewDataSource {
             let viewcontroller = NewCategoryViewController()
-            viewcontroller.onAddCategoryButtonTapped = { subtitleNameCategory in
-                self.addSubtitleToCategory(subtitleNameCategory)
+            viewcontroller.onAddCategoryButtonTapped = { trackerCategoryName in
```
И тут

Aleksandr0406 replied:
добавил везде, где возвращал значение через кожуру(на экране создания привычки/нерегулярного события, на экране выбора что создать привычку или нерегулярное событие и на главный экран

Andreynnt commented on Tracker/TrackersViewController/NewTrackerViewController.swift:
```diff
@@ -74,20 +74,20 @@ final class NewTrackerViewController: UIViewController {
-            addNewNotRegularTrackerButton.leadingAnchor.constraint(equalTo: view.leadingAnchor, constant: 20),
-            addNewNotRegularTrackerButton.trailingAnchor.constraint(equalTo: view.trailingAnchor, constant: -20),
-            addNewNotRegularTrackerButton.heightAnchor.constraint(equalToConstant: 60)
+            addNewNotRegularEventButton.topAnchor.constraint(equalTo: addNewTrackerButton.bottomAnchor, constant: 16),
```
Константы часто выносят в приватный enum/struct в этом же файле:

```
private enum Constants {
    static let someViewHeight: CGFloat = 10
    static let someViewCornerRadius: CGFloat = 8
    ...
}
```
Советую делать также, чтобы все цифры лежали в одном месте и не было "magic numbers"

Aleksandr0406 replied:
а какие цифры лучше выносить, все или есть какая то логика? Например на экране выбора что создать привычку или нерегулярное событие(NewTrackerViewController) есть цифры: размер шрифта текста в кнопках, радиусы скрепления  кнопок, значения констрейнтов. Вроде бы по месту понятно что каждое из значений означает или выносить лучше, чтобы потом легче менять?

Aleksandr0406 committed:
style: исправить названия файлов на подходящие

Названия файлов, связанных с созданием экрана создания привычки или нерегулярного события, имели неподходящие названия, указывающие как будто только на создание привычки

Aleksandr0406 committed:
refactor: исправить force unwrap в проекте

Aleksandr0406 committed:
refactor:  перенести создание JSONDecoder/Encoder в свойство соответствующего класса

Aleksandr0406 committed:
refactor: исправить нэйминг файлов на подходящие

Aleksandr0406 committed:
refactor: исправить нэйминг файла на подходящий, избавиться от force unwrap

Aleksandr0406 committed:
refactor: удаление retain cycle'ов

Aleksandr0406 committed:
style: переименовать названии папки создания и трекера/нерегулярки

Aleksandr0406 committed:
style: исправить шрифты и сепараторы

Везде добавлен необходимый weight шрифтов по фигме. Во всех таблицах убран сепаратор у самой нижней ячейки, остальные сепараторы выровнены в соответствии с фигмой

Aleksandr0406 committed:
refactor: убрать retain cycle и поправить шрифты

Экран выбора созданной категории

Aleksandr0406 committed:
style: исправить стиль кнопки добавления трекера и поправить шрифты

Главный экран приложения

Aleksandr0406 commented:
@Andreynnt привет, спасибо за ревью!

1. Force unwrap'ы и force cast'ы поправил кроме мест, где есть  try! self.init(context: context). Выше в комменте написал более полно

2. Retain cycle-ы убрал везде, где нашел

3. Дизайн исправил в соответсвии с фигмой

Ссылку направил, буду ждать обратной связи.
